From 745fa9005b7c99d4540b91d6a1934fbe5425c25b Mon Sep 17 00:00:00 2001
From: Christian Zietz <czietz@gmx.net>
Date: Tue, 26 Dec 2017 11:40:47 +0000
Subject: [PATCH 2/3] Make it explicit that ide_get_data/ide_put_data only
 transfer whole sectors, allowing an easier unrolling of the loops

---
 bios/ide.c | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/bios/ide.c b/bios/ide.c
index b87b6c3..c89db38 100644
--- a/bios/ide.c
+++ b/bios/ide.c
@@ -720,12 +720,13 @@ static void ide_get_data_32(volatile struct IDE *interface,UBYTE *buffer,ULONG b
 /*
  * get data from IDE device
  */
-static void ide_get_data(volatile struct IDE *interface,UBYTE *buffer,ULONG bufferlen,int need_byteswap)
+static void ide_get_data(volatile struct IDE *interface,UBYTE *buffer,UWORD numsecs,int need_byteswap)
 {
+    ULONG bufferlen = (ULONG)numsecs * SECTOR_SIZE;
     XFERWIDTH *p = (XFERWIDTH *)buffer;
     XFERWIDTH *end = (XFERWIDTH *)(buffer + bufferlen);
 
-    KDEBUG(("ide_get_data(%p, %p, %lu, %d)\n", interface, buffer, bufferlen, need_byteswap));
+    KDEBUG(("ide_get_data(%p, %p, %u, %d)\n", interface, buffer, numsecs, need_byteswap));
 
 #if CONF_WITH_APOLLO_68080
     if (is_apollo_68080)
@@ -799,9 +800,9 @@ static LONG ide_read(UBYTE cmd,UWORD ifnum,UWORD dev,ULONG sector,UWORD count,UB
         rc = E_OK;
         if (status1 & IDE_STATUS_DRQ) {
             if (info->twisted_cable) {
-                ide_get_data((volatile struct IDE *)(((ULONG)interface)+1),buffer,xferlen,need_byteswap);
+                ide_get_data((volatile struct IDE *)(((ULONG)interface)+1),buffer,numsecs,need_byteswap);
             } else {
-                ide_get_data(interface,buffer,xferlen,need_byteswap);
+                ide_get_data(interface,buffer,numsecs,need_byteswap);
             }
         } else {
             rc = EREADF;
@@ -828,8 +829,9 @@ static LONG ide_read(UBYTE cmd,UWORD ifnum,UWORD dev,ULONG sector,UWORD count,UB
 /*
  * send data to IDE device
  */
-static void ide_put_data(volatile struct IDE *interface,UBYTE *buffer,ULONG bufferlen,int need_byteswap)
+static void ide_put_data(volatile struct IDE *interface,UBYTE *buffer,UWORD numsecs,int need_byteswap)
 {
+    ULONG bufferlen = (ULONG)numsecs * SECTOR_SIZE;
     XFERWIDTH *p = (XFERWIDTH *)buffer;
     XFERWIDTH *end = (XFERWIDTH *)(buffer + bufferlen);
 
@@ -894,9 +896,9 @@ static LONG ide_write(UBYTE cmd,UWORD ifnum,UWORD dev,ULONG sector,UWORD count,U
         status1 = IDE_READ_STATUS();    /* status, clear pending interrupt */
         if (status1 & IDE_STATUS_DRQ) {
             if (info->twisted_cable) {
-                ide_put_data((volatile struct IDE *)(((ULONG)interface)+1),buffer,xferlen,need_byteswap);
+                ide_put_data((volatile struct IDE *)(((ULONG)interface)+1),buffer,numsecs,need_byteswap);
             } else {
-                ide_put_data(interface,buffer,xferlen,need_byteswap);
+                ide_put_data(interface,buffer,numsecs,need_byteswap);
             }
         } else {
             rc = EWRITF;
-- 
2.8.2.windows.1

